### ISSUE TITLE
Replace glsl_to_spirv in examples

### ISSUE THREAD

**Issue by LTheGreats** [gfx-rs/gfx#2546]
Short info header:
- GFX version:
- OS:
- GPU:

**Comment by Michael-Lfx** [comment]
glsl-to-spirv is also used in vulkan backend.

**Comment by LTheGreats** [comment]
Really this isn't just about the examples, but all of the gfx-rs code that depends on the deprecated library. (Sorry about the many name changes.)

**Comment by grovesNL** [comment]
I think it should be removed from the Vulkan backend already, see https://github.com/gfx-rs/gfx/pull/2533

**Comment by kvark** [comment]
To clarify: GLSL is in now way a part of gfx-hal API, so the issue is rather harmless. It just affects the way people start using the project, since they replicate the examples, and we shouldn't promote a deprecated library there. So, a documentation issue for the most part.

**Comment by aleksijuvani** [comment]
I don't think shaderc-rs is a very good alternative. It's kind of a pain to build on Windows, requiring a C++ compiler (e.g. usually Visual Studio, which is an ordeal to install), Python and CMake in the path. glsl_to_spirv used glslang instead and included a pre-built binary for Windows.

I'm sure it's not a problem for experienced developers, but it won't make for a good experience for anyone looking to learn Rust or gfx-hal and doesn't have these already installed.

**Comment by kvark** [comment]
Good point! Maybe we can just undeprecate in then?

**Comment by Michael-Lfx** [comment]
> I don't think shaderc-rs is a very good alternative. It's kind of a pain to build on Windows, requiring a C++ compiler (e.g. usually Visual Studio, which is an ordeal to install), Python and CMake in the path. glsl_to_spirv used glslang instead and included a pre-built binary for Windows.
> 
> I'm sure it's not a problem for experienced developers, but it won't make for a good experience for anyone looking to learn Rust or gfx-hal and doesn't have these already installed.

IIRC shaderc is based on glslang, spriv-cross and other KHR projects. And yes it requires a complex environment to build which is difficult for beginners. Here we just need glslang to convert glsl to Vulkan compatible spirv binary. I think it's good to make it simple.  Because [TimNN/glslang-sys](https://github.com/TimNN/glslang-sys) stays in May 10, 2015, I have a proposal that gfx should hold a glslang-rs crate to follow the upstream. @kvark

**Comment by kvark** [comment]
> I have a proposal that gfx should hold a glslang-rs crate to follow the upstream

I don't think we, the gfx-rs team, should be owning it. As I mentioned above, our entry point is SPIR-V and we care about all the conversions *from* it. The real reason for us to depend on glslang (in examples only!) is about documentation, since just providing raw SPV shaders would not be very beginner friendly or readable.

**Comment by Michael-Lfx** [comment]
Got it. We have to rely on glslang or something like it to do the realtime shader conversion actually. So we are seeking a update-to-date glslang-sys crate just like Josh maintains the spirv-cross-sys crate.

**Comment by grovesNL** [comment]
I started work on a wrapper for glslang a few months ago so we could use it from Rust directly instead of having to call into a separate process through glslang's CLI. I think it would be slightly more performant and we could think about the API design a bit more and map types direectly.

Unfortunately as far as I could tell glslang is focused on CLI usage more than usage as a library. If I recall correctly, there seemed to be quite a bit logic remapping the CLI options to the C++ options, but it's too bad because the CLI seems to have a pretty good API.

So to experiment with it, I changed its `main` function slightly (and renamed it from `main` to anything else) so I can invoke it easily from Rust and build a wrapper around it, like I did with SPIRV-Cross. After I did all of this I noticed efforts like shaderc-rs etc. and decided it wasn't worth it for me to maintain it because I'm busy enough already 😆 If there's demand for this and somebody willing to maintain it, I'd be happy to assist with setting it up the same way as spirv_cross.

**Comment by grovesNL** [comment]
Also I think the only reason glsl_to_spirv may avoid the compiler on Windows is because it distributes a pre-built binary on Windows, but I don't think this is a very good solution. I haven't had many issues with people building spirv_cross through cc-rs on Windows (besides sometimes needing to configure CI a bit), so maybe people are hitting a different problem with shaderc, or there is a bug somewhere?

**Comment by aleksijuvani** [comment]
Unless you're using the Vulkan backend I suppose, which does not depend on spirv_cross.

**Comment by kvark** [comment]
Sounds like shaderrc is the winner for now.

**Comment by Lokathor** [comment]
I'll note here that Rust on windows already demands VS Studio for the msvc toolchain, so it's reasonably likely to already be on the dev's machine. And if you're using the gnu toolchain, well you only do that for compatibility with other gnu libs, which strongly implies you've got a gnu compiler installed.

There might be any other concern, but "needs a c++ compiler" isn't a big one if you're already doing rust at all.

**Comment by aleksijuvani** [comment]
Interesting. I suppose they need that for the linker. I had no idea.

**Comment by breakin** [comment]
I'm just now starting looking at Rust and gfx-rs but I am experienced windows programmer. Whatever you end up doing please make sure that Cargo build/run just works for the examples without any manual steps. Having MSVC tollchain installed is ok, but no manual steps like "place dlls here and there" or "build this solution and then do x and z" or "use mingw" or such. Thanks!

**Comment by aleksijuvani** [comment]
We could be able to avoid the need to install CMake if we wrapped glslang like @grovesNL explained above, but instead of using CMake to generate the glslang project files, manually listed the files required for the build. This way we would only require MSVC to be installed, and everything should work out of the box.
